### PR TITLE
Allow ObjectParsers to specify required sets of fields

### LINKED ISSUE #48824: Aggregations should throw parse exception if no field or script is specified
Today if you specify an aggregation with no `field` or `script` like so:

```json
{
  "size": 0,
  "aggs": {
    "the_terms": {
      "terms": {
        "size": 10
      }
    }
  }
}
```

We throw an exception:

```json
{
    "type": "illegal_state_exception",
    "reason": "value source config is invalid; must have either a field context or a script or marked as unwrapped"
}
```

The problem is that the exception is generated from what is an otherwise invalid ValuesSourceConfig object.  E.g. when resolving we identify that there is no script or field, [generate an invalid VSConfig](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java#L74-L79), then later at the factory layer check to see if the VSConfig is valid and [throw an exception](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java#L255-L258).

Besides being confusing from a code perspective, it also means we don't throw until we've already broadcast the request to the cluster.  We should instead fail during parsing because there is no sensible reason to omit both field and script (I don't think?)

Note: There are a few instances (e.g. `ParentAggregationBuilder`) that use the mutability of VSConfig which may be why we ended up in this situation.  But even if we don't fix that part of VSConfig, we can head off these scenarios at the parse layer first. 

Related to https://github.com/elastic/elasticsearch/issues/42949

### PR DESCRIPTION
**Pull request by polyfractal**
`ConstructingObjectParser` can be used to specify required fields, but it is still difficult to configure "sets" of fields where only one of the set is required (`(foo OR bar) AND (bizz OR buzz)`).  It requires hand-rolled logic in each `ConstructingObjectParser`, or special validation methods that are called during/after object construction.

This commit adds a new method on `ObjectParser` which allows the parsers to register required sets.  E.g. `["foo", "bar"]` can be registered, which means "foo", "bar" or both must be configured by the user otherwise an exception is thrown.

This pattern crops up in many places in our parsers, but as a demonstration this PR adds validation for aggregation `"field"` and `"script"` fields.  One or both must be configured on all aggregations, and omitting both should result in an exception.  This was previously handled far downstream resulting in an aggregation exception, when it should be a parse exception.

This might be a controversial change given it's partial overlap with `ConstructingObjectParser`.  Open to changes/suggestions, as well as better naming :)

Closes https://github.com/elastic/elasticsearch/issues/48824

### CODE REVIEW

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/ObjectParser.java [hunk 1]
```diff
@@ -235,11 +236,32 @@ public Value parse(XContentParser parser, Value value, Context context) throws I
...
                     fieldParser.assertSupports(name, parser, currentFieldName);
+
+                    // Check to see if this field is a required field, if it is we can
+                    // remove the entry as the requirement is satisfied
+                    Iterator<String[]> iter = requiredFieldSets.iterator();
+                    while (iter.hasNext()) {
+                        String[] requriedFields = iter.next();
+                        for (String field : requriedFields) {
+                            if (field.equals(currentFieldName)) {
+                                iter.remove();
```
Comment (rjernst) [inline libs/x-content/src/main/java/org/elasticsearch/common/xcontent/ObjectParser.java]: Isn't this removing the value from the ObjectParser's instance, which is static?

File: server/src/test/java/org/elasticsearch/search/aggregations/AggregatorFactoriesBuilderTests.java [hunk 2]
```diff
@@ -139,13 +139,13 @@ private static AggregationBuilder getRandomAggregation() {
         final int randomAggregatorPoolSize = 4;
         switch (randomIntBetween(1, randomAggregatorPoolSize)) {
         case 1:
-            return AggregationBuilders.avg(randomAlphaOfLengthBetween(3, 10));
+            return AggregationBuilders.avg(randomAlphaOfLengthBetween(3, 10)).field("foo");
```
Comment (rjernst) [inline server/src/test/java/org/elasticsearch/search/aggregations/AggregatorFactoriesBuilderTests.java]: These test changes seem unrelated?

File: docs/reference/aggregations/bucket/range-aggregation.asciidoc [hunk 3]
```diff
@@ -379,30 +379,3 @@ Response:
 }
 --------------------------------------------------
 // TESTRESPONSE[s/\.\.\.//]
-
-If a sub aggregation is also based on the same value source as the range aggregation (like the `stats` aggregation in the example above) it is possible to leave out the value source definition for it. The following will return the same response as above:
```
Comment (rjernst) [inline docs/reference/aggregations/bucket/range-aggregation.asciidoc]: This doc change seems unrelated to this PR which is about adding infrastructure?

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/AbstractObjectParser.java [hunk 4]
```diff
@@ -211,6 +213,27 @@ public void declareIntArray(BiConsumer<Value, List<Integer>> consumer, ParseFiel
...
+     * specified by the user.  If neither of those fields are configured, an exception will be thrown.
+     *
+     * Multiple required sets can be configured:
+     *
+     *   declareRequiredFieldSet(new String[]{"foo", "bar"});
+     *   declareRequiredFieldSet(new String[]{"bizz", "buzz"});
+     *
+     * requires that ("foo" OR "bar") AND ("bizz" OR "buzz") are configured by the user
+     */
+    public void declareRequiredFieldSet(String[] requriedSet) {
```
Comment (rjernst) [inline libs/x-content/src/main/java/org/elasticsearch/common/xcontent/AbstractObjectParser.java]: This could be a varargs so the call to the method doesn't require explicitly creating an array?

**Review by rjernst — COMMENTED** [review]
Overall having better control over what combinations of fields are allowed or must be present, but I wonder if this is sufficient. The examples given in javadoc are pretty simplistic, and shown like a query, while we are talking about json objects. Would it be possible to give closer to a real world example?

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/ObjectParser.java (hunk 1 above)
Comment (polyfractal) [reply libs/x-content/src/main/java/org/elasticsearch/common/xcontent/ObjectParser.java]: Oof, you're right.  I was just looking at the class itself (which is not static) but everyone declares ObjectParser as static.  I'll fiddle to get rid of the mutability

**Comment by polyfractal** [comment]
Thanks!

>  The examples given in javadoc are pretty simplistic, and shown like a query, while we are talking about json objects. Would it be possible to give closer to a real world example?

The motivating example for me was from aggregations: `script` or `field` (or both) _must_ be specified, but if both are omitted it should be a parse exception.  That's also why the PR has agg-related test fixes and documentation tweaks, since the PR included this as a demonstration example.I'll adjust the javadocs to show some json examples instead of boolean logic.

Regarding the agg-specific parts of the PR, I can remove them if we want it to be just infrastructure, and followup with a second PR for the agg stuff.

**Comment by polyfractal** [comment]
I backed out all the agg-related changes (making `field`/`script` required, the test tweaks and the documentation change) so this PR is now just related to the object parser.

Fixed access to the static object by making a copy of the list before parsing.  This didn't feel especially clever, so open to alternatives :)

Also beefed up the javadocs a bit with json examples

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/AbstractObjectParser.java [hunk 5]
```diff
@@ -211,6 +213,49 @@ public void declareIntArray(BiConsumer<Value, List<Integer>> consumer, ParseFiel
         declareField(consumer, (p, c) -> parseArray(p, () -> itemParser.parse(p, c)), field, type);
     }
 
+    /**
+     * Declares a set of fields that are required for parsing to succeed. Only one of the values
+     * provided per String[] must be matched.  E.g.
+     *
+     * declareRequiredFieldSet(new String[]{"foo", "bar"}); means "foo" or "bar" (or both) fields must be
```
Comment (rjernst) [inline libs/x-content/src/main/java/org/elasticsearch/common/xcontent/AbstractObjectParser.java]: Instead of "or both", maybe say "at least one of"?

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/AbstractObjectParser.java [hunk 6]
```diff
@@ -211,6 +213,49 @@ public void declareIntArray(BiConsumer<Value, List<Integer>> consumer, ParseFiel
...
+     * specified by the user.  If neither of those fields are configured, an exception will be thrown.
```
Comment (rjernst) [inline libs/x-content/src/main/java/org/elasticsearch/common/xcontent/AbstractObjectParser.java]: configured -> present

File: libs/x-content/src/test/java/org/elasticsearch/common/xcontent/ObjectParserTests.java [hunk 7]
```diff
@@ -753,4 +755,182 @@ public void testConsumeUnknownFields() throws IOException {
...
+                this.c = value;
+            }
+
+            private void setD(long value) {
+                this.d = value;
+            }
+        }
+
+        {
+            XContentParser parser = createParser(JsonXContent.jsonXContent, "{\"unrelated\": \"123\"}");
```
Comment (rjernst) [inline libs/x-content/src/test/java/org/elasticsearch/common/xcontent/ObjectParserTests.java]: Could these blocks be individual test methods? Teh test structs could be uniquely named and exist at the root of the class instead of locally within this method?

**Review by rjernst — COMMENTED** [review]
A few more comments.

**Comment by polyfractal** [comment]
Review comments addressed!  The extra javadoc markup made the json snippets significantly less readable though so not sure if they are still helpful or not...

I opted to ignore `<p>` since none of the other javadocs in this class use it either.  I also used `<code>` instead of `{@code}` to avoid escaping all the braces

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/AbstractObjectParser.java [hunk 8]
```diff
@@ -211,6 +213,60 @@ public void declareIntArray(BiConsumer<Value, List<Integer>> consumer, ParseFiel
...
+     *   <tr><td><code>{"foo":"..."}</code></td><td>Missing "bizz" or "buzz" field</td></tr>
+     *   <tr><td><code>{"bar":"..."}</code></td><td>Missing "bizz" or "buzz" field</td></tr>
+     *   <tr><td><code>{"bizz": "..."}</code></td><td>Missing "foo" or "bar" field</td></tr>
+     *   <tr><td><code>{"buzz": "..."}</code></td><td>Missing "foo" or "bar" field</td></tr>
+     *   <tr><td><code>{"foo":"...", "bar": "..."}</code></td><td>Missing "bizz" or "buzz" field</td></tr>
+     *   <tr><td><code>{"bizz":"...", "buzz": "..."}</code></td><td>Missing "foo" or "bar" field</td></tr>
+     *   <tr><td><code>{"unrelated":"..."}</code></td>  <td>Missing "foo" or "bar" field, and missing "bizz" or "buzz" field</td></tr>
+     * </table>
+     *
+     * @param requriedSet
```
Comment (rjernst) [inline libs/x-content/src/main/java/org/elasticsearch/common/xcontent/AbstractObjectParser.java]: typo: requriedSet -> requiredSet

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/AbstractObjectParser.java [hunk 9]
```diff
@@ -211,6 +213,60 @@ public void declareIntArray(BiConsumer<Value, List<Integer>> consumer, ParseFiel
...
+     *          A set of required fields, where at least one of the fields in the array _must_ be present
+     */
+    public void declareRequiredFieldSet(String... requriedSet) {
```
Comment (rjernst) [inline libs/x-content/src/main/java/org/elasticsearch/common/xcontent/AbstractObjectParser.java]: typo in the arg name too

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by polyfractal** [comment]
Had a chance to revisit this, think I resolved the javadoc compilation build issue.  Unsure _why_ the "summary" field broke everything, but replacing it with a `<content>` tag instead satisfies the build. (kudos to @mark-vieira for the tip).

This got stale, merged in master and fixed conflicts, will monitor to make sure there aren't any new test issues before merging.

**Comment by polyfractal** [comment]
@elasticmachine update branch
